internal/core/service/user_service.go

```diff
@@ -0,0 +1,160 @@
+package service
+
+import (
+	"errors"
+	"github.com/gin-gonic/gin"
+	"github.com/viniciusgferreira/ps-tag-onboarding-go/internal/core/domain/model"
+	"github.com/viniciusgferreira/ps-tag-onboarding-go/internal/core/port"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"regexp"
+)
+
+var (
+	ErrUserNotFound      = errors.New("User not found")
+	ErrUserAlreadyExists = errors.New("User with the same first and last name already exists")
+	ErrInvalidAge        = errors.New("User must be at least 18 years old")
+	ErrInvalidName       = errors.New("User first and last name cannot be empty")
+	ErrInvalidEmail      = errors.New("Invalid email format")
+	ErrInvalidID         = errors.New("Invalid id (ObjectID)")
+)
+
+type UserService struct {
+	repo port.UserRepository
+}
+
+func New(repo port.UserRepository) *UserService {
+	return &UserService{repo: repo}
+}
+
+type CustomError struct {
+	Message string   `json:"error"`
+	Details []string `json:"details,omitempty"`
+}
+
+func NewCustomError(err error) CustomError {
+	return CustomError{Message: err.Error()}
+}
+
+func NewValidationError(validationErrors []error) CustomError {
+	details := make([]string, len(validationErrors))
+	for i, err := range validationErrors {
+		details[i] = err.Error()
+	}
+	return CustomError{
+		Message: "User did not pass validation",
+		Details: details,
+	}
+}
+
+func (r CustomError) Error() string {
+	return r.Message
+}
+
+func (s *UserService) Find(ctx *gin.Context, id string) (*model.User, error) {
+	err := s.validateID(id)
+	if err != nil {
+		return nil, err
+	}
+	user, err := s.repo.FindById(ctx, id)
+	if err != nil {
+		return nil, NewCustomError(err)
+	}
+	if user == nil {
+		return nil, NewCustomError(ErrUserNotFound)
+	}
+	return user, nil
+}
+
+func (s *UserService) Save(ctx *gin.Context, u model.User) (*model.User, error) {
+	validationErrors := s.Validate(u)
+	if validationErrors != nil {
+		return nil, NewValidationError(validationErrors)
+	}
+	exists, err := s.repo.ExistsByFirstNameAndLastName(ctx, u.FirstName, u.LastName)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return nil, NewValidationError([]error{ErrUserAlreadyExists})
+	}
+	user, err := s.repo.Save(ctx, u)
+	if err != nil {
+		return nil, err
+	}
+	return user, nil
+}
+
+func (s *UserService) Update(ctx *gin.Context, u model.User) (*model.User, error) {
+	err := s.validateID(u.ID)
+	if err != nil {
+		return nil, err
+	}
+	validationErrors := s.Validate(u)
+	if validationErrors != nil {
+		return nil, NewValidationError(validationErrors)
+	}
+	_, err = s.repo.FindById(ctx, u.ID)
+	if err != nil {
+		return nil, NewCustomError(ErrUserNotFound)
+	}
```

Comment:
If the user doesn't exists you could return 404

Reply:
I think 404 code is related to the handler/httpserver which is doing that (returning 404), returning a custom error 
 at the service layer can help to test and also better readability.

Reply:
I agree! What I mean is to validate if FindById returns a nil for the user. And if there was an error returned in FindById it should return a 500 error (that should be checked in httpserver package)